gateway/src/shard/processor/impl.rs

```diff
@@ -34,8 +35,184 @@ use twilight_model::gateway::{
         identify::{Identify, IdentifyInfo, IdentifyProperties},
         resume::Resume,
     },
+    GatewayIntents,
 };
-use url::Url;
+use url::{ParseError as UrlParseError, Url};
+
+#[cfg(not(feature = "simd-json"))]
+use serde_json::Error as JsonError;
+#[cfg(feature = "simd-json")]
+use simd_json::Error as JsonError;
+
+/// Connecting to the gateway failed.
+#[derive(Debug)]
+pub enum ConnectingError {
+    Establishing { source: TungsteniteError },
+    ParsingUrl { source: UrlParseError, url: String },
+}
+
+impl Display for ConnectingError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        match self {
+            Self::Establishing { source } => Display::fmt(source, f),
+            Self::ParsingUrl { source, url } => f.write_fmt(format_args!(
+                "the gateway url `{}` is invalid: {}",
+                url, source,
+            )),
+        }
+    }
+}
+
+impl Error for ConnectingError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        match self {
+            Self::Establishing { source } => Some(source),
+            Self::ParsingUrl { source, .. } => Some(source),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum GatewayEventParsingError {
+    /// Deserializing the GatewayEvent payload from JSON failed.
+    Deserializing {
+        /// Reason for the error.
+        source: JsonError,
+    },
+    /// The payload received from Discord was an invalid structure.
```

Comment:
This error would probably happen most often if Discord changed their JSON structures, so maybe having a note that we didn't know what was, but that it *may* still be correct on Discord's end could be useful. 

```suggestion
    /// The payload received from Discord was not a recognized or valid structure.
```

Reply:
addressed